// requiresafe/lib/get-help.js
var fs = require('fs');
var path = require('path');
var dir = __dirname + '/commands';
var marked = require('marked');
var chalk = require('chalk');
var TerminalRenderer = require('marked-terminal');


marked.setOptions({
    // Define custom renderer 
    renderer: new TerminalRenderer({
        em: chalk.blue,
        heading: chalk.blue.bold,
        codespan: chalk.blue
    })
});

var indent = function (str, space) {
    return str.split('\n').map(function (line) {
        return space + line;
    }).join('\n');
};

module.exports = fs.readdirSync(dir)
    .filter(function (fileName) {
        return fileName !== 'index.js';
    })
    .map(function (file) {
        var contents = fs.readFileSync(dir + '/' + file, 'utf8');
        var commandName = path.basename(file, '.js');

        // extract first comment for help contents
        var help = contents.split('*/', 1)[0].split('\n').slice(1).slice(0, -1).join('\n');
        var header = chalk.magenta('requiresafe ' + commandName);
        var hasHelp = contents.slice(0, 2) === '/*';

        if (!hasHelp) {
            help = '(none)\n';
        }

        help = marked(help).split('\n').slice(0, -1).join('\n');

        header = indent(header, '  ');
        help = indent(help, '  ');

        return {
            help: header + '\n' + help,
            name: commandName
        };
    });
