// requiresafe/lib/commands/login.js
/*
Prompts for username and password. When validated, this will store an auth token in your `$HOME/.requiresaferc` config.
*/
var inquirer = require('inquirer');
var chalk = require('chalk');
var updateConfig = require('../update-config');


function required(value) {
    return !!value.trim() || chalk.red('required');
}


module.exports = function (config, callback) {
    var questions = [
        {
            name: 'email',
            message: 'enter your email',
            validate: required,
            default: config.email
        },
        {
            name: 'password',
            message: 'enter your password (never saved)',
            validate: required,
            type: 'password'
        }
    ];

    inquirer.prompt(questions, function (answers) {
        // attempt to save username to config for future use
        updateConfig({email: answers.email}, function (err) {
            
            if (err) {
                console.log(err);
            }

            config.api.login({
                email: answers.email,
                password: answers.password,
                expires: false,
            }, function (err, data) {
                if (err) {
                    console.log('error', err);
                } else {
                    var token = data.authToken;
                    updateConfig({token: token}, function () {
                        if (err) {
                            return callback(err);
                        } else {
                            return callback(null, 'You\'re logged in. An auth token has been saved to $HOME/.requiresaferc');
                        }
                    });
                }
            });
        });
    });
};
